source/distributed/tria_base.cc

```diff
@@ -164,9 +164,14 @@ namespace parallel
 
     {
       // find ghost owners
-      for (const auto &cell : this->active_cell_iterators())
-        if (cell->is_ghost())
-          number_cache.ghost_owners.insert(cell->subdomain_id());
+      {
+        const auto vertices_with_ghost_neighbors =
+          GridTools::compute_vertices_with_ghost_neighbors(*this);
+        for (const auto &vertex_with_ghost_neighbors :
+             vertices_with_ghost_neighbors)
+          for (const auto &ghost_neigbor : vertex_with_ghost_neighbors.second)
+            number_cache.ghost_owners.insert(ghost_neigbor);
```

Comment:
But I don't understand why you need this. If you had a `p::s::T` that corresponds to a long string of cells, subdivided into three partitions, then on processor 0 the only immediately neighboring ghost cells belong to processor 1, but you also have ghost cells from processor 2 -- they just don't share a vertex with the locally owned cells. Why do you not want to consider them here? These are still ghost cells, so they should be reflected in the ghost owners, right?

If the issue is that there are *users* of the `number_cache.ghost_owners` set that only want to send stuff to immediate neighbors, then that's a separate issue -- in that case, we need a separate field -- but this seems like the wrong place to address this.

Reply:
I agree, we can keep `ghost_owners` as it is and introduce a new vector (maybe: `direct_ghost_owners`).

One question: normally, both are identical; exception: p:s:t without artificial cells. What is the right place to fill `direct_ghost_owners` for latter. Here, in `parallel::TriangulationBase` or in `p:s:t`?

Reply:
Another place where you'd get a difference is if we ever have a triangulation that allows us to have more than one layer of ghost cells.

If you can provide the functionality in a base class for all derived classes, then that's worth it. It might cost a bit more time for derived classes where this would be cheap to compute (p::d::T right now), but this is computed at refinement time once. The cost is almost certainly completely negligible to everything else we do after mesh refinement.